opm/grid/CpGrid.hpp

```diff
@@ -1024,6 +1024,13 @@ namespace Dune
                                            const std::vector<std::array<int,3>>& cells_per_dim_vec,
                                            int preAdaptMaxLevel);
 
+        void updateCornerHistoryLevels(const std::vector<std::vector<std::array<int,2>>>& cornerInMarkedElemWithEquivRefinedCorner,
+                                       std::map<std::array<int,2>,std::array<int,2>> elemLgrAndElemLgrCorner_to_refinedLevelAndRefinedCorner,
+                                       std::unordered_map<int,std::array<int,2>> adaptedCorner_to_elemLgrAndElemLgrCorner,
```

Comment:
I can double-check, but I remembered trying to pass argument with const & (once they existed, i.e., previously populated), but in the case of (unordered) maps, I was getting errors. I do not remember the error messages. Maybe something related to hash/(not)being-hashable. 

I can add the cont & and see which error appears, if any.

Reply:
My memory was kind of in the right direction, in the sense that I did try not to copy but there were errors so I removed the const references from the arguments. 

When I try to pass the map as a const reference instead of a copy, I get:

error: passing ‘const std::map<std::array<int, 2>, std::array<int, 2> >’ as ‘this’ argument discards qualifiers [-fpermissive]

...

usr/include/c++/12/bits/stl_map.h:502:7: note:   in call to ‘std::map<_Key, _Tp, _Compare, _Alloc>::mapped_type& std::map<_Key, _Tp, _Compare, _Alloc>::operator[](const key_type&) [with _Key = std::array<int, 2>; _Tp = std::array<int, 2>; _Compare = std::less<std::array<int, 2> >; _Alloc = std::allocator<std::pair<const std::array<int, 2>, std::array<int, 2> > >; mapped_type = std::array<int, 2>; key_type = std::array<int, 2>]’
  502 |       operator[](const key_type& __k)

Reply:
You cannot use operator[] on a const map (or const map reference), because it might insert a new object. You can use at() instead (throws instead of inserting a default-constructed item if lookup fails for a key), or find() for more fine-grained control (returns end iterator if not found).

Reply:
The reason for this error is that `operator[]` in maps is not there as a const method, because if there is no entry for the key in a map, then such an entry will be automatically inserted. This requires a mutable objects. That is probably not what you want here

Hence methods that get a const map may never use `operator[]`, but always need to use either `at(key)` which will return the value if it is there or throw or `find(key)` which will return an iterator (that might either be the end iterator or point to another entry) that you need to check.

So to avoid bugs it is always a good idea to use the `at` function that has a const version instead. It will throw an exception if there is no entry for the key:
```
const auto& val = map.at(key);
```
or use find and do the error checking yourself (cumbersome):
```
auto candidate = map.find(key);
assert (candidate != map.end && candidate->first == key); // use an assert if we are 100% sure that this value should exist.
const auto& val = candidate->second;
```

Reply:
Thanks @atgeirr and @blattms for the explanation!